src/state/opamRepositoryState.ml

```diff
@@ -171,27 +171,32 @@ let load lock_kind gt =
   let repositories = OpamRepositoryName.Map.mapi mk_repo repos_map in
   let repos_tmp_root = lazy (OpamFilename.mk_tmp_dir ()) in
   let repos_tmp = Hashtbl.create 23 in
-  OpamRepositoryName.Map.iter (fun name repo ->
-      let uncompressed_root = OpamRepositoryPath.root gt.root repo.repo_name in
-      let tar = OpamRepositoryPath.tar gt.root repo.repo_name in
-      if not (OpamFilename.exists_dir uncompressed_root) &&
-         OpamFilename.exists tar
-      then
-        let tmp = lazy (
-          let tmp_root = Lazy.force repos_tmp_root in
-          try
-            (* We rely on this path pattern to clean the repo.
-               cf. [clean_repo_tmp] *)
-            OpamFilename.extract_in tar tmp_root;
-            OpamFilename.Op.(tmp_root / OpamRepositoryName.to_string name)
-          with Failure s ->
-            OpamFilename.remove tar;
-            OpamConsole.error_and_exit `Aborted
-              "%s.\nRun `opam update --repositories %s` to fix the issue"
-              s (OpamRepositoryName.to_string name);
-        ) in
-        Hashtbl.add repos_tmp name tmp
-    ) repositories;
+  if OpamRepositoryConfig.(!r.repo_tarring) then
+    OpamRepositoryName.Map.iter (fun name repo ->
+        let uncompressed_root = OpamRepositoryPath.root gt.root repo.repo_name in
+        let tar = OpamRepositoryPath.tar gt.root repo.repo_name in
+        if not (OpamFilename.exists_dir uncompressed_root) &&
+           OpamFilename.exists tar
+        then
+          let tmp = lazy (
+            let tmp_root = Lazy.force repos_tmp_root in
+            try
+              (* We rely on this path pattern to clean the repo.
+                 cf. [clean_repo_tmp] *)
+              OpamFilename.extract_in tar tmp_root;
+              OpamFilename.Op.(tmp_root / OpamRepositoryName.to_string name)
+            with Failure s ->
+              OpamFilename.remove tar;
+              OpamConsole.error_and_exit `Aborted
+                "%s.\nRun `opam update --repositories %s` to fix the issue"
+                s (OpamRepositoryName.to_string name))
+          in
+          Hashtbl.add repos_tmp name tmp)
+      repositories
+  else
+    OpamRepositoryName.Map.iter (fun _name repo ->
+        OpamFilename.remove (OpamRepositoryPath.tar gt.root repo.repo_name))
```

Comment:
I’m a bit surprised to see the remove being in `lock` instead of `OpamUpdate`.
I would expect only `opam update` to actually do anything.

I haven’t tried locally yet but unless I’ve missed something, I would expect such code to make opam crash for users calling `opam install <pkg>` just after having upgraded from opam 2.1 because they would have to get the content of the repository but it wouldn’t be available anywhere anymore.

But it’s just a hunch from reading the code, I’ll try locally later.

Reply:
Yep it’s making it fail:
```
opam@feac57c1cc94:~/opam$ opam-2.1 install ocamlfind
The following actions will be performed:
  - install ocamlfind 1.9.2

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
-> retrieved ocamlfind.1.9.2  (http://download.camlcity.org/download/findlib-1.9.2.tar.gz)
-> installed ocamlfind.1.9.2
Done.
opam@feac57c1cc94:~/opam$ ls ~/.opam/repo
beta  default  lock  repos-config  state-0287C1DB.cache
opam@feac57c1cc94:~/opam$ opam-2.1 update
This development version of opam requires an update to the layout of /home/opam/.opam from version 2.0 to version 2.1, which can't be reverted.
You may want to back it up before going further.

Continue? [Y/n] y
Format upgrade done.

<><> Updating package repositories ><><><><><><><><><><><><><><><><><><><><><><>
[beta] no changes from git+https://github.com/ocaml/ocaml-beta-repository
[default] synchronised from file:///home/opam/opam-repository

<><> Synchronising development packages <><><><><><><><><><><><><><><><><><><><>
opam@feac57c1cc94:~/opam$ ls ~/.opam/repo
beta.tar.gz  default.tar.gz  lock  repos-config  state-0287C1DB.cache
opam@feac57c1cc94:~/opam$ opam-2.1 remove ocamlfind
The following actions will be performed:
  - remove ocamlfind 1.9.2

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
-> removed   ocamlfind.1.9.2
Done.
opam@feac57c1cc94:~/opam$ ls ~/.opam/repo
beta.tar.gz  default.tar.gz  lock  repos-config  state-0287C1DB.cache
opam@feac57c1cc94:~/opam$ ./opam install ocamlfind
[ERROR] No package named ocamlfind found.
opam@feac57c1cc94:~/opam$ ls ~/.opam/repo
lock  repos-config  state-0287C1DB.cache  state-030998F6.cache
```
I believe it’s also making the opam-rt testsuite fail

Reply:
Indeed, It is the wrong place